### ISSUE TITLE
Flow is warning me for path on Windows that are ignored

### ISSUE THREAD

**Issue by MoOx** [facebook/flow#2187]
I have added `.*/node_modules/.*` in the `[ignore]` section, but flow is warning me about some path. Is this on purpose? I expect ignored path to be completely ignored, even for warnings.

```
On Windows, paths must be less than 248 characters for directories and 260 characters for files. This path has 265 characters. Skipping C:\xxxxxx (my project path) xxxxxx\node_modules\nyc\node_modules\istanbul\node_modules\handlebars\node_modules\uglify-js\node_modules\yargs\node_modules\cliui\node_modules\center-align\node_modules\align-text\node_modules\kind-of\node_modules\is-buffer
```

**Comment by MoOx** [comment]
The weird part: when I just run "flow" (not "flow check") I don't have the warnings (but still "Found 0 errors").
[Per doc](https://flowtype.org/docs/cli.html), I understand that "flow" is checking the current folder (so in my case my root) and so it seems to be the same as "flow check" (but it's not, since the latest "flow check" give me warnings that the first "flow" does not).

**Comment by thejameskyle** [comment]
cc @gabelevi 

Side note: It looks like you are using npm 2, you might be able to use npm 3 to get rid of the warning (just completely blow away node_modules and reinstall).

**Comment by MoOx** [comment]
@thejameskyle sadly, I am using npm 3!

**Comment by gabelevi** [comment]
I think you probably started off using npm2 and then upgraded to npm3. This doesn't seem to remove node_modules, so you might want to try

```
rm -rf node_modules
npm3 install # or npm install if npm3 is globally installed
```

The "On Windows, paths must be less than 248 characters for directories..." message is a warning, and won't actually affect the running of Flow, except that paths that are too long are ignored. The warning is generated before ignores are processed, when we're just trying to find the files in the first place.

**Comment by MoOx** [comment]
FYI, I never used npm 2 on my windows machine since it was a fresh setup.

**Comment by gabelevi** [comment]
@MoOx are you sure? The warning mentions the path

```
node_modules\nyc\node_modules\istanbul\node_modules\handlebars\node_modules\uglify-js\node_modules\yargs\node_modules\cliui\node_modules\center-align\node_modules\align-text\node_modules\kind-of\node_modules\is-buffer
```

which has lots of nested node_modules. As I understand it, npm3 automatically flattens, so I don't know how you'd end up with this directory structure unless it was created by npm2. Or is the directory structure coming from somewhere else? Like is this somehow checked into your repo?

Does deleting `node_modules`, and running `npm install` fix the directory structure?

**Comment by MoOx** [comment]
I am sure, fresh "rm -rf node_modules && npm install".
npm 3 "try to" flatten "when possible", so if you have some packages that require differents versions of some packages, it won't flatten thoses deps. That's the reason.

**Comment by gabelevi** [comment]
TL;DR Either I'm missing something or `npm3 install`/`npm3 dedupe` is kind of broken and that really sucks for Windows.

Something is really screwy with npm3. I tried something like

```Bash
$ mkdir foo
$ cd foo
$ mkdir node_modules
$ npm install --global npm3
$ npm3 --version
3.10.6
$ npm3 install nyc@^6.4.3 # peaked at phenomic's package.json
```

What I expected was for npm3 to install the package and automatically dedupe. However, what I see is this:

```Bash
$ ls node_modules/
nyc
```

Why is there only one top level npm dependency? And there is a path that is 21 directories deep?!

```Bash
$ find . -type d -exec bash -c 'echo $(tr -cd / <<< "$1"|wc -c):$1' -- {} \;  | sort -n | tail -n 1 | awk -F: '{print $1, $2}'
21 ./node_modules/nyc/node_modules/istanbul/node_modules/handlebars/node_modules/uglify-js/node_modules/yargs/node_modules/cliui/node_modules/right-align/node_modules/align-text/node_modules/kind-of/node_modules/is-buffer/test
```

Furthermore, shouldn't `npm3 dedupe` be a no-op? However running it actually changes things immensely:

```Bash
$ npm3 dedupe
$ ls -1 node_modules/ | wc -l
     151
$ find . -type d -exec bash -c 'echo $(tr -cd / <<< "$1"|wc -c):$1' -- {} \;  | sort -n | tail -n 1 | awk -F: '{print $1, $2}'
15 ./node_modules/uglify-js/node_modules/yargs/node_modules/cliui/node_modules/right-align/node_modules/align-text/node_modules/kind-of/node_modules/is-buffer/test
```

Ok, a depth of 15 is better, but still really huge! Are there really that many dependency conflicts? Let's check out `cliui`

```Bash
$ for dir in `find . -name "cliui" -type d`; do grep -H "_id" $dir/package.json; done
./node_modules/cliui/package.json:  "_id": "cliui@3.2.0",
./node_modules/uglify-js/node_modules/yargs/node_modules/cliui/package.json:  "_id": "cliui@2.1.0",
```

Ok, so we have cliui@3.2.0 installed at level 1 and cliui@2.1.0 installed at level 3. Why isn't `npm3 install` or `npm3 dedupe` installing cliui@2.1.0 at level 2? Maybe something at level 2 needs to require `cliui@3.2.0`, so moving 2.1.0 from level 3 to level 2 would mess that up?

```Bash
$ find . -name "package.json" | xargs grep '"cliui":'
./node_modules/uglify-js/node_modules/yargs/package.json:    "cliui": "^2.1.0",
./node_modules/yargs/package.json:    "cliui": "^3.2.0",
```

Nope, that doesn't seem to be it. Maybe cliui@2.1.0 requires something at level 3, and moving it up would mess that up? Nope, all it's dependencies are

```JavaScript
  "dependencies": {
    "center-align": "^0.1.1",
    "right-align": "^0.1.1",
    "wordwrap": "0.0.2"
  },
```

and those are all in its `node_modules`:

```Bash
$ ls node_modules/uglify-js/node_modules/yargs/node_modules/cliui/node_modules/
center-align right-align  wordwrap
```

**Comment by MoOx** [comment]
I have to admit this is totally confusing... As a workaround, can't we check paths and emit the warning AFTER removing the ignored files? Most people will probably ignore node_modules, and so won't have any warnings ^^

**Comment by gabelevi** [comment]
It's harder than it sounds, I'm afraid. Due to this limitation of Windows, we can't complete our traversal of the file system. This means the traversal might stop at a directory, rather than make it all way to the files under that directory.

So what you're asking is for us to try and figure out whether a given directory is included and not ignored.

Includes are globs, so they might look like `foo/**/bar/*.js`. So given `foo/dir`, you can't say definitively whether its included or not.

Ignores are regexes, intended to match against the file paths. So given `foo/.*/test.js`, it's hard to say whether `foo/dir` is ignored.

So we could do a best effort - checking only the beginnings of globs and ignoring ignore regexes that don't match. However, there is a cost to doing these extra checks, especially iterating through the list of regexes. This cost is already slowing down the parsing system, since we do these checks for every file we find. I'd rather not do additional checks for each directory we find.

That said, when we emit a warning, we probably could do a best-effort check of the ignore regexes against the file or directory causing the warning. That seems reasonable.

**Comment by raajnadar** [comment]
This issue is still not fixed??

Node v10
Npm v6
